igvc_navigation/src/field_d_planner/main.cpp

```diff
@@ -0,0 +1,268 @@
+  int lookahead_dist;          // number of cell traversals to look ahead at when decising next position along path
+  float occupancy_threshold;   // maximum occupancy probability before a cell is considered to have infinite traversal
+                               // cost
+  // publish path for path_follower
+  ros::Publisher path_pub = nh.advertise<nav_msgs::Path>("/path", 1);
+
+  igvc::getParam(pNh, "c_space", configuration_space);
+  igvc::getParam(pNh, "maximum_distance", maximum_distance);
+  igvc::getParam(pNh, "rate", rate_time);
+  igvc::getParam(pNh, "goal_range", goal_range);
+  igvc::getParam(pNh, "publish_expanded", publish_expanded);
+  igvc::getParam(pNh, "follow_old_path", follow_old_path);
+  igvc::getParam(pNh, "lookahead_dist", lookahead_dist);
+  igvc::getParam(pNh, "occupancy_threshold", occupancy_threshold);
+
+  planner.NodeGrid.setConfigurationSpace(static_cast<float>(configuration_space));
+  planner.NodeGrid.setOccupancyThreshold(static_cast<float>(occupancy_threshold));
+  planner.setGoalDistance(static_cast<float>(goal_range));
+
+  ros::Rate rate(rate_time);  // path update rate
+
+  int num_nodes_updated = 0;
+  int num_nodes_expanded = 0;
+
+  bool initialize_search = true;  // set to true if the search problem must be initialized
+
+  while (ros::ok())
+  {
+    ros::spinOnce();  // handle subscriber callbacks
+
+    // don't plan unless the map has been initialized and a goal node has been set
+    if (initialize_graph)
+      continue;
+    else
+      planner.NodeGrid.updateGraph(map);
+
+    // don't plan unless a goal node has been set
+    if (!initial_goal_set)
+      continue;
+
+    if (initialize_search)
+      planner.initializeSearch();
+
+    if (goal_changed)
+    {
+      ROS_INFO_STREAM("New Goal Received. Initializing Search...");
+      initialize_search = true;
+      goal_changed = false;
+      continue;
+    }
+
+    // gather cells with updated edge costs and update affected nodes
+    num_nodes_updated = planner.updateNodesAroundUpdatedCells();
+
+    if (num_nodes_updated > 0)
+      ROS_INFO_STREAM(num_nodes_updated << " nodes updated");
+
+    // only update the graph if nodes have been updated
+    if ((num_nodes_updated > 0) || initialize_search)
+    {
+      ros::Time begin = ros::Time::now();
+      num_nodes_expanded = planner.computeShortestPath();
+
+      double elapsed = (ros::Time::now() - begin).toSec();
+      ROS_INFO_STREAM(num_nodes_expanded << " nodes expanded in " << elapsed << "s.");
+      if (initialize_search)
+        initialize_search = false;
+    }
+
+    planner.constructOptimalPath(lookahead_dist);
+
+    if (publish_expanded)
+      publish_expanded_set(planner.getExplored(), expanded_cloud, expanded_pub);
+
+    nav_msgs::Path path_msg;
+    path_msg.header.stamp = ros::Time::now();
+    path_msg.header.frame_id = "odom";
+
+    for (Position pos : planner.Path)
```

Comment:
Move this to a separate method? this while loop is very long.

Reply:
I might create a wrapper class which takes in a `NodeHandle` and runs the path planner. Thoughts on this? Similar to [this](https://github.com/RoboJackets/igvc-software/blob/feat/refactor_smooth_control/igvc_navigation/src/path_follower/Path_follower.cpp)

Reply:
Yeah, I think that'd be a nice refactor, to separate the ROS stuff from the actual algorithm.